applications/CompressiblePotentialFlowApplication/custom_processes/define_2d_wake_process.cpp

```diff
@@ -0,0 +1,326 @@
+        const int thread = OpenMPUtils::ThisThread();
+        auto& thread_trailing_edge_element_ids = all_threads_trailing_edge_element_ids[thread];
+
+        // Check if the element is touching the trailing edge
+        CheckIfTrailingEdgeElement(it_elem, thread_trailing_edge_element_ids);
+
+        // Elements downstream the trailing edge can be wake elements
+        potentially_wake = CheckIfPotentiallyWakeElement(it_elem);
+
+        if (potentially_wake) {
+            // Compute the nodal distances of the element to the wake
+            nodal_distances_to_wake = ComputeNodalDistancesToWake(it_elem);
+
+            // Selecting the cut (wake) elements
+            is_wake_element = CheckIfWakeElement(nodal_distances_to_wake);
+
+            // Mark wake element and save their nodal distances to the wake
+            if (is_wake_element) {
+                it_elem->SetValue(WAKE, true);
+                it_elem->SetValue(ELEMENTAL_DISTANCES, nodal_distances_to_wake);
+                auto r_geometry = it_elem->GetGeometry();
+                for (unsigned int i = 0; i < it_elem->GetGeometry().size(); i++) {
+                    r_geometry[i].SetLock();
+                    r_geometry[i].FastGetSolutionStepValue(DISTANCE) = nodal_distances_to_wake(i);
+                    r_geometry[i].UnSetLock();
+                }
+            }
+        }
+    }
+    // Add the trailing edge elements to the trailing_edge_sub_model_part
+    AddTrailingEdgeElements(number_of_threads, all_threads_trailing_edge_element_ids);
+}
+
+// This function checks if the element is touching the trailing edge
+void Define2DWakeProcess::CheckIfTrailingEdgeElement(const ElementIteratorType& rElement,
+                                                    std::unordered_set<std::size_t>& thread_trailing_edge_element_ids)
+{
+    // Loop over element nodes
+    for (unsigned int i = 0; i < rElement->GetGeometry().size(); i++) {
+        // Elements touching the trailing edge are trailing edge elements
+        if (rElement->GetGeometry()[i].Id() == mTrailingEdgeNode->Id()) {
+            rElement->SetValue(TRAILING_EDGE, true);
+            thread_trailing_edge_element_ids.insert(rElement->Id());
+        }
+    }
+}
+
+// This function selects the elements downstream the trailing edge as
+// potentially wake elements
+const bool Define2DWakeProcess::CheckIfPotentiallyWakeElement(const ElementIteratorType& rElement)
+{
+    // Compute the distance from the trailing edge to the element's center
+    const auto distance_to_element_center =
+        ComputeDistanceFromTrailingEdgeToPoint(rElement->GetGeometry().Center());
+
+    // Compute the projection of the distance in the wake direction
+    return inner_prod(distance_to_element_center, mWakeDirection) > 0.0 ? true : false;
+}
+
+// This function computes the distance of the element nodes to the wake
+const BoundedVector<double, 3> Define2DWakeProcess::ComputeNodalDistancesToWake(const ElementIteratorType& rElement)
```

Comment:
```suggestion
const BoundedVector<double, 3> Define2DWakeProcess::ComputeNodalDistancesToWake(const Element& rElement)
```

I don't think you need the iterator